typing/typecore.ml

```diff
@@ -1749,748 +1749,23 @@ and is_nonexpansive_opt = function
-            (Env.empty, Use.empty)
-            bindings
-  and value_binding : Env.env -> Typedtree.value_binding -> Env.env * Use.t =
-    (* NB: returns new environment only *)
-    fun env { vb_pat; vb_expr } ->
-      let vars = pattern_variables vb_pat in
-      let ty = expression env vb_expr in
-      let ty = if is_destructuring_pattern vb_pat then Use.inspect ty else ty in
-      (List.fold_left
-        (fun env id -> Ident.add id ty env)
-        Env.empty
-        vars,
-       ty)
-  and is_destructuring_pattern : Typedtree.pattern -> bool =
-    fun pat -> match pat.pat_desc with
-      | Tpat_any -> false
-      | Tpat_var (_, _) -> false
-      | Tpat_alias (pat, _, _) -> is_destructuring_pattern pat
-      | Tpat_constant _ -> true
-      | Tpat_tuple _ -> true
-      | Tpat_construct (_, _, _) -> true
-      | Tpat_variant _ -> true
-      | Tpat_record (_, _) -> true
-      | Tpat_array _ -> true
-      | Tpat_or (l,r,_) -> is_destructuring_pattern l || is_destructuring_pattern r
-      | Tpat_lazy _ -> true
-
-  let check_recursive_expression env idlist expr =
-    let ty = expression (build_unguarded_env idlist) expr in
-    match Use.unguarded ty, Use.dependent ty, classify_expression expr with
-    | _ :: _, _, _ (* The expression inspects rec-bound variables *)
-    | _, _ :: _, Dynamic -> (* The expression depends on rec-bound variables
-                               and its size is unknown *)
-        raise(Error(expr.exp_loc, env, Illegal_letrec_expr))
-    | [], _, Static (* The expression has known size *)
-    | [], [], Dynamic -> (* The expression has unknown size,
-                            but does not depend on rec-bound variables *)
-        ()
-  let check_class_expr env idlist ce =
-    let rec class_expr : Env.env -> Typedtree.class_expr -> Use.t =
-      fun env ce -> match ce.cl_desc with
-        | Tcl_ident (_, _, _) -> Use.empty
-        | Tcl_structure _ -> Use.empty
-        | Tcl_fun (_, _, _, _, _) -> Use.empty
-        | Tcl_apply (_, _) -> Use.empty
-        | Tcl_let (rec_flag, valbinds, _, ce) ->
-            let _, ty = value_bindings rec_flag env valbinds in
-            Use.join ty (class_expr env ce)
-        | Tcl_constraint (ce, _, _, _, _) ->
-            class_expr env ce
-        | Tcl_open (_, _, _, _, ce) ->
-            class_expr env ce
-    in
-    match Use.unguarded (class_expr (build_unguarded_env idlist) ce) with
-    | [] -> ()
-    | _ :: _ -> raise(Error(ce.cl_loc, env, Illegal_class_expr))
-end
-
 let check_recursive_bindings env valbinds =
   let ids = List.concat
-      (List.map (fun b -> pattern_variables b.vb_pat) valbinds) in
+      (List.map (fun b -> pat_bound_idents b.vb_pat) valbinds) in
   List.iter
     (fun {vb_expr} ->
-       Rec_check.check_recursive_expression env ids vb_expr)
+       try Rec_check.check_recursive_expression ids vb_expr
+       with Rec_check.Illegal_expr ->
+         raise(Error(vb_expr.exp_loc, env, Illegal_letrec_expr))
```

Comment:
I'm not super-fond of exceptions crossing module boundaries¹ if we can avoid it. Would it make sense to expose `is_valid_{recursive_expression,class_express}` functions that return a boolean, or some `option` thing, that we could use here?

¹: I think it's ok to expose functions in your .mli that may raise an exception (and document it), but I would prefer to also have a recommended API for the main entry points that does not rely on exceptional return for a core aspect of its behavior, unless there is strong justification for why that is more convenient.

Reply:
I agree with you.

I hadn't done the change because I wanted this PR to be a trivial "code move, no semantic change" PR. It is done now.